### ISSUE TITLE
CanBeNonNullable false-negative

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#4492]
## Expected Behavior

`assignedExpression` violates `CanBeNonNullable`.

```kotlin
private fun visitAssignment(assignedExpression: KtExpression?) {
    if (assignedExpression == null) return
    val name = if (assignedExpression is KtQualifiedExpression) {
        assignedExpression.selectorExpression
    } else {
        assignedExpression
    }?.text ?: return
    assignments.getOrPut(name) { mutableSetOf() }.add(assignedExpression)
}
```

## Observed Behavior
detekt is fine with that code

## Context
If a function returns `Unit` any null check to exit the function if the parameter is `null` should be consider as a parameter that could be non-null.

I saw this false-negative while working on this PR: #4491

## Your Environment
* Version of detekt used: main
* Link to your project (if it's a public repository): detekt.

**Comment by cortinico** [comment]
Isn't #4431 supposed to catch this actually?

**Comment by BraisGabin** [comment]
Yes, but it is not.

**Comment by VitalyVPinchuk** [comment]
Hi!
Can I work on this?

**Comment by BraisGabin** [comment]
Please go ahead :)

**Comment by VitalyVPinchuk** [comment]
Should `!is` in `if` / `when` statements also be reported?
e.g.
```
fun foo(a: Int?) {
    if (a !is Int) return
    println(a)
}
```
